Formal2024/Workshops/workshop8.lean

```diff
@@ -0,0 +1,312 @@
+    applied by the `symm` tactic.
+
+  The most important one of these is the `@[simp]` attribute. It makes the marked lemma available
+  to the *simplifier*, which can be invoked with the `simp` tactic. In this course, `simp` is
+  disabled by the `Library` file, so we haven't imported it in this file.
+-/
+
+/-
+  The *simplifier* is aware of hundreds (if not thousands) of `@[simp]`-marked lemmas that it can
+  use to rewrite an expression into something (hopefully) simpler. Its power stems from the fact
+  that you can add your own lemmas to the simplifier, so it becomes aware of your own definitions.
+
+  As with `rw` you can provide lemmas to the `simp` tactic manually when you use it, by using `[]`.
+  However, if you find that you are using the same lemmas over and over, there is a good chance
+  they deserve to me marked with the `@[simp]` attribute.
+
+  You can use the `simp?` tactic to see what lemmas `simp` is using. You can also use the `#simp`
+  command to see how `simp` would simplify a given expression.
+-/
+
+section simp
+
+variable {G : Type} [Group G] (x : G)
+
+example : x * 1 * x⁻¹ = x⁻¹ * 1 * 1 * x := by simp?
+-- The lemmas on the output list are all marked with `@[simp]`.
+
+#simp x * (1 * x) * x⁻¹
+
+/-
+  We can't just make any lemma a `simp` lemma. Its conclusion needs to be an equality or iff.
+  Moreover, for a lemma to be useful to `simp`, the RHS of this equality should be 'simpler' than
+  the LHS. Of course, this is just a heuristic, but it is usually good enough.
+-/
+
+-- For example, this lemma is a good candidate for a `simp` lemma.
+example (n : ℕ) : n + 0 = n := by simp?
+
+-- But this one is not.
+lemma not_simp (n : ℕ) : n = n + 0 := by simp?
+
+-- If both of these were `simp` lemmas, `simp` could get stuck in an infinite loop.
+-- Let's test this by making the bad lemma a `simp` lemma locally.
+attribute [local simp] not_simp
+
+-- The following example will give a maximum recursion reached error because of this.
+-- example (n : ℕ) : n = 0 + n := by simp
+
+end simp
+
+
+/-
+  ## Defining your own notation
+
+  To use symbols such as `+`, `*`, `Σ`, `Π` in Lean, they need to be declared as custom notation.
+  This can be done using the `notation` command and its various incarnations.
+-/
+
+-- This is how one can define binary, prefix and postfix operators.
+infixl:65   " + " => HAdd.hAdd  -- left-associative
+infix:50    " = " => Eq         -- non-associative
+infixr:80   " ^ " => HPow.hPow  -- right-associative
+prefix:100  "-"   => Neg.neg
+postfix:max "⁻¹"  => Inv.inv
+
+-- Note that we can only assign notation to functions that are already defined.
+
+-- The previous lines are equivalent to these, which use the full `notation` command.
+notation:65 lhs:65 " + " rhs:66 => HAdd.hAdd lhs rhs
+notation:50 lhs:51 " = " rhs:51 => Eq lhs rhs
+notation:80 lhs:81 " ^ " rhs:80 => HPow.hPow lhs rhs
+notation:100 "-" arg:100 => Neg.neg arg
+notation:1024 arg:1024 "⁻¹" => Inv.inv arg
```

Comment:
This last bit feels a bit rushed. Maybe we could skip the notation command and just do a couple of examples of new notation, showing the difference of left/right associativity?

Reply:
Just added some more examples, and explained the precedence number and how it gives left and right associativity.